process/block/alteredaccounts/alteredAccountsProvider.go

```diff
@@ -0,0 +1,348 @@
+	}
+
+	return &alteredAccountsProvider{
+		shardCoordinator: args.ShardCoordinator,
+		addressConverter: args.AddressConverter,
+		accountsDB:       args.AccountsDB,
+		marshalizer:      args.Marshalizer,
+		fungibleTokensIdentifier: map[string]struct{}{
+			core.BuiltInFunctionESDTTransfer:         {},
+			core.BuiltInFunctionESDTBurn:             {},
+			core.BuiltInFunctionESDTLocalMint:        {},
+			core.BuiltInFunctionESDTLocalBurn:        {},
+			core.BuiltInFunctionESDTWipe:             {},
+			core.BuiltInFunctionMultiESDTNFTTransfer: {},
+		},
+		nonFungibleTokensIdentifier: map[string]struct{}{
+			core.BuiltInFunctionESDTNFTTransfer:      {},
+			core.BuiltInFunctionESDTNFTBurn:          {},
+			core.BuiltInFunctionESDTNFTAddQuantity:   {},
+			core.BuiltInFunctionESDTNFTCreate:        {},
+			core.BuiltInFunctionMultiESDTNFTTransfer: {},
+		},
+	}, nil
+}
+
+// ExtractAlteredAccountsFromPool will extract and return altered accounts from the pool
+func (aap *alteredAccountsProvider) ExtractAlteredAccountsFromPool(txPool *indexer.Pool) (map[string]*indexer.AlteredAccount, error) {
+	markedAccounts := make(map[string]*markedAlteredAccount)
+	aap.extractMoveBalanceAddresses(txPool, markedAccounts)
+	err := aap.extractESDTAccounts(txPool, markedAccounts)
+	if err != nil {
+		return nil, err
+	}
+
+	return aap.fetchDataForMarkedAccounts(markedAccounts)
+}
+
+func (aap *alteredAccountsProvider) fetchDataForMarkedAccounts(markedAccounts map[string]*markedAlteredAccount) (map[string]*indexer.AlteredAccount, error) {
+	alteredAccounts := make(map[string]*indexer.AlteredAccount)
+	var err error
+	for address, markedAccount := range markedAccounts {
+		err = aap.processMarkedAccountData(address, markedAccount.tokens, alteredAccounts)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return alteredAccounts, nil
+}
+
+func (aap *alteredAccountsProvider) processMarkedAccountData(
+	encodedAddress string,
+	markedAccountTokens []*markedAlteredAccountToken,
+	alteredAccounts map[string]*indexer.AlteredAccount,
+) error {
+	addressBytes, err := aap.addressConverter.Decode(encodedAddress)
+	if err != nil {
+		log.Warn("cannot decode marked altered account address", "error", err)
+		return err
+	}
+
+	account, err := aap.accountsDB.LoadAccount(addressBytes)
+	if err != nil {
+		log.Warn("cannot load account when computing altered accounts",
+			"address", encodedAddress,
+			"error", err)
+		return err
+	}
+
+	userAccount, ok := account.(state.UserAccountHandler)
+	if !ok {
+		log.Warn("cannot cast AccountHandler to UserAccountHandler", "address", encodedAddress)
+		return err
+	}
+
+	alteredAccounts[encodedAddress] = &indexer.AlteredAccount{
+		Address: encodedAddress,
+		Balance: userAccount.GetBalance().String(),
+		Nonce:   userAccount.GetNonce(),
+	}
+
+	if len(markedAccountTokens) > 0 {
+		for _, tokenData := range markedAccountTokens {
+			err = aap.fetchTokensDataForMarkedAccount(encodedAddress, tokenData, alteredAccounts)
```

Comment:
Why not pass `userAccount` instead of `encodedAddress`?
This way, in `fetchTokensDataForMarkedAccount` you would have saved **3 duplicated code checks**:
1. Decode address
2. Load account from DB
3. Convert account to user account

Reply:
:+1:

Reply:
true. sent user account as well. kept the encodedAddress because it tells me directly the map key